extensions/spark/kyuubi-extension-spark-3-5/src/test/scala/org/apache/spark/sql/KyuubiSparkSQLExtensionTest.scala

```diff
@@ -99,23 +102,49 @@ trait KyuubiSparkSQLExtensionTest extends QueryTest
     withListener(sql(sqlString))(callback)
   }
 
-  def withListener(df: => DataFrame)(callback: DataWritingCommand => Unit): Unit = {
+  def withListener(df: => DataFrame)(
+      callback: DataWritingCommand => Unit,
+      failIfNotCallback: Boolean = true): Unit = {
+    val writes = Collections.synchronizedList(new java.util.ArrayList[DataWritingCommand]())
+
     val listener = new QueryExecutionListener {
       override def onFailure(f: String, qe: QueryExecution, e: Exception): Unit = {}
 
       override def onSuccess(funcName: String, qe: QueryExecution, duration: Long): Unit = {
-        qe.executedPlan match {
-          case write: DataWritingCommandExec => callback(write.cmd)
-          case _ =>
+        def collectWrite(plan: SparkPlan): Unit = {
+          plan match {
+            case write: DataWritingCommandExec =>
+              writes.add(write.cmd)
+            case a: AdaptiveSparkPlanExec => collectWrite(a.executedPlan)
+            case _ =>
+          }
         }
+        collectWrite(qe.executedPlan)
       }
     }
+    // Make sure the listener is registered after all previous events have been processed
+    sparkContext.listenerBus.waitUntilEmpty()
     spark.listenerManager.register(listener)
     try {
       df.collect()
       sparkContext.listenerBus.waitUntilEmpty()
     } finally {
       spark.listenerManager.unregister(listener)
     }
+    if (failIfNotCallback && writes.isEmpty) {
+      fail("No write command found")
+    }
+    writes.forEach(callback(_))
+  }
+
+  def collectRebalancePartitions(plan: LogicalPlan): Seq[RebalancePartitions] = {
+    def collect(p: LogicalPlan): Seq[RebalancePartitions] = {
+      p.flatMap {
+        case r: RebalancePartitions => Seq(r)
+        case s: LogicalQueryStage => collect(s.logicalPlan)
```

Comment:
why do we need to match `LogicalQueryStage` ?

Reply:
https://github.com/apache/kyuubi/blob/b4838b40e6a9074697918ccecd3dfe71cc52442d/extensions/spark/kyuubi-extension-spark-3-3/src/test/scala/org/apache/spark/sql/RebalanceBeforeWritingSuite.scala#L51

This test case gets a logical plan like this:

![image](https://github.com/user-attachments/assets/6f0683bb-27a7-40c7-8a23-ecbb5d3a55ba)

Reply:
I guess it's because the logical plan is obtained by `write.cmd`

https://github.com/apache/kyuubi/blob/b4838b40e6a9074697918ccecd3dfe71cc52442d/extensions/spark/kyuubi-extension-spark-3-5/src/test/scala/org/apache/spark/sql/KyuubiSparkSQLExtensionTest.scala#L108